src/lib.rs

```diff
@@ -862,41 +864,189 @@ unsafe impl<'a> alloc::Alloc for &'a Bump {
         let old_size = layout.size();
 
         if new_size <= old_size {
-            // Shrinking an allocation is easy: Just give the same pointer back
-            // If it's the last allocation, we can reclaim some of the bytes, otherwise
-            // they are simply leaked.
-            if self.is_last_allocation(ptr, layout) {
-                let new_end = NonNull::new_unchecked(ptr.as_ptr().add(new_size));
-                self.current_chunk_footer.get().as_ref().ptr.set(new_end);
+            if self.is_last_allocation(ptr)
+                // Only reclaim the excess space (which requires a copy) if it
+                // is worth it: we are actually going to recover "enough" space
+                // and we can do a non-overlapping copy.
+                && new_size <= old_size / 2
+            {
+                let delta = old_size - new_size;
+                let footer = self.current_chunk_footer.get();
+                let footer = footer.as_ref();
+                footer
+                    .ptr
+                    .set(NonNull::new_unchecked(footer.ptr.get().as_ptr().add(delta)));
+                let new_ptr = footer.ptr.get();
+                // NB: we know it is non-overlapping because of the size check
+                // in the `if` condition.
+                ptr::copy_nonoverlapping(ptr.as_ptr(), new_ptr.as_ptr(), new_size);
```

Comment:
Have you tested how much we loose by using `ptr::copy` instead and now having the `new_size <= old_size / 2` check?

Reply:
Ah, I see. If we are shrinking but not by a lot, we can just use the same pointer. We **could** reclaim it, but it is a lot of trouble for very few bytes saved. I agree with this implementation! :+1:

Reply:
Exactly, and since we already are doing this calculus, we might was well choose the threshold where we get to do a faster copy as well. That said, if you want to experiment with other implementations and benchmark them, I'm happy to accept results-driven PRs! :)